File: externals/smc/lib/ObjC/statemap.m
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is State Machine Compiler (SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2007. Charles W. Rapp.
// All Rights Reserved.
//
// Contributor(s):
//    Chris Liscio
//
// Namespace
//	statemap
//
// Description
//  This namespace contains the finite state machine context
//  class. The user can derive FSM contexts from this class and
//  interface to them with the methods of this class.
//
// Notes
//  The finite state machine needs to be initialized to the
//  starting state of the FSM.  This must be done manually in
//  the constructor of the derived class. 
//
// Author
//	Chris Liscio
//
// RCS ID
// Id: statemap.m,v 1.5 2013/07/14 14:32:36 cwrapp Exp
//
// CHANGE LOG
// (See bottom of this file)
//

#import "statemap.h"

@implementation SMCState
- (id)initWithName:(NSString*)name stateId:(int)stateId;
{
    self = [super init];
    if (!self) {
        return nil;
    }
    
    _name = [name copy];
    _stateId = stateId;
    
    return self;
}
- (int)stateId
{
    return _stateId;
}
- (void)setStateId:(int)newStateId
{
    _stateId = newStateId;
}
- (NSString*)name
{
    return _name;
}

- (void)setName:(NSString*)aValue
{
    NSString* oldName = _name;
    _name = [aValue S_RETAIN];
    [oldName S_RELEASE];
}

- (void)dealloc
{
    [_name S_RELEASE];
    [super S_DEALLOC];
}
@end

@implementation SMCStateEntry
+ (id)stateEntryWithState:(SMCState*)state next:(SMCStateEntry*)next;
{
    SMCStateEntry *ret = [[[SMCStateEntry alloc] init] S_AUTORELEASE];
    [ret setState:state];
    [ret setNext:next];
    return ret;
}

- (void)dealloc
{
    [self setState:nil];
    [self setNext:nil];
    [super S_DEALLOC];
}

- (SMCState*)state;
{
    return _state;
}

- (void)setState:(SMCState*)aValue;
{
    id old = _state;
    _state = [aValue S_RETAIN];
    [old S_RELEASE];
}

- (SMCStateEntry*)next;
{
    return _next;
}

- (void)setNext:(SMCStateEntry*)aValue;
{
    id old = _next;
    _next = [aValue S_RETAIN];
    [old S_RELEASE];
}
@end

@implementation SMCFSMContext
- (id) initWithState: (SMCState*) aState;
{
    self = [super init];

    if (!self)
    {
        return nil;
    }

    _state = [aState S_RETAIN];

    return self;
}

- (BOOL)isInTransition;
{
    return(_state == NULL ? YES : NO);
}

- (NSString*)transition
{
    return _transition;
}

- (void)setTransition:(NSString*)aValue
{
    NSString* oldTransition = _transition;
    _transition = [aValue S_RETAIN];
    [oldTransition S_RELEASE];
}

- (void)dealloc
{
    [_previousState S_RELEASE];
    [_state S_RELEASE];
    while (_stateStack != NULL) {
        SMCStateEntry *entry = _stateStack;
        _stateStack = [_stateStack next];
        [entry S_RELEASE];
    }
    [super S_DEALLOC];
}

- (void)clearState
{
    [_previousState S_RELEASE]; _previousState = [_state S_RETAIN];
    [_state S_RELEASE]; _state = NULL;
}

- (SMCState*)previousState;
{
    return _previousState;
}

- (void)setState:(SMCState*)state;
{
    if (state != _state) {
        [_state S_RELEASE];
        _state = [state S_RETAIN];
        if ([self debugFlag]) {
            TRACE( @"ENTER STATE     : %@\n\r", [_state name] );
        }
    }
}

- (BOOL)isStateStackEmpty;
{
    return (_stateStack == nil);
}

// Returns the state stack's depth.
- (int)stateStackDepth
{
    SMCStateEntry *state_ptr;
    int retval;

    for (state_ptr = _stateStack, retval = 0;
         state_ptr != NULL;
         state_ptr = [state_ptr next], ++retval)
        ;

    return (retval);
}

// Push the current state on top of the state stack
// and make the specified state the current state.
- (void)pushState:(SMCState*)state
{
    SMCStateEntry *new_entry;

    // Do the push only if there is a state to be pushed
    // on the stack.
    if (_state != NULL)
    {
        new_entry = [[SMCStateEntry stateEntryWithState:_state next:_stateStack] S_RETAIN];
        _stateStack = new_entry;
    }

    [_state S_RELEASE]; _state = [state S_RETAIN];

    if ([self debugFlag]) {
        TRACE(@"PUSH TO STATE   : %@\n\r", [_state name]);
    }
}

// Make the state on top of the state stack the
// current state.
- (void)popState
{
    SMCStateEntry *entry;

    // Popping when there was no previous push is an error.
    NSAssert(_stateStack != NULL, @"Popping empty state stack");

    [_state S_RELEASE]; _state = [[_stateStack state] S_RETAIN];
    entry = _stateStack;
    _stateStack = [_stateStack next];
    [entry S_RELEASE];

    if ([self debugFlag]) {
        TRACE(@"POP TO STATE    : %@\n\r", [_state name]);
    }
}

// Remove all states from the state stack.
- (void)emptyStateStack
{
    SMCStateEntry *state_ptr, *next_ptr;

    for (state_ptr = _stateStack;
         state_ptr != NULL;
         state_ptr = next_ptr)
    {
        next_ptr = [state_ptr next];
        [state_ptr S_RELEASE];
    }
    
    _stateStack = NULL;
}

- (BOOL)debugFlag
{
    return _debugFlag;
}

- (void)setDebugFlag:(BOOL)newDebugFlag
{
    _debugFlag = newDebugFlag;
}

@end

//
// CHANGE LOG
// Log: statemap.m,v
// Revision 1.5  2013/07/14 14:32:36  cwrapp
// check in for release 6.2.0
//
// Revision 1.4  2011/11/20 14:58:33  cwrapp
// Check in for SMC v. 6.1.0
//
// Revision 1.3  2009/11/24 20:42:39  cwrapp
// v. 6.0.1 update
//
// Revision 1.2  2009/04/10 14:02:10  cwrapp
// Set initial state via initializer.
//
// Revision 1.1  2007/01/15 00:23:50  cwrapp
// Release 4.4.0 initial commit.
//

File: externals/smc/lib/Ruby/statemap.rb
#
# The contents of this file are subject to the Mozilla Public
# License Version 1.1 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is State Machine Compiler (SMC).
#
# The Initial Developer of the Original Code is Charles W. Rapp.
#
# Port to Ruby by Francois Perrad, francois.perrad@gadz.org
# Copyright 2004, Francois Perrad.
# All Rights Reserved.
#
# Contributor(s):
#
# RCS ID
# Id: statemap.rb,v 1.5 2010/09/11 18:55:55 fperrad Exp
#

# This namespace contains two class (and two exceptions) :
#* State
#    the base State class
#* FSMContext
#    the Finite State Machine Context class
#
#See: http://smc.sourceforge.net/
module Statemap

	# A StateUndefinedException is thrown by
	# an SMC-generated state machine whenever a transition is taken
	# and there is no state currently set. This occurs when a
	# transition is issued from with a transition action.
	class StateUndefinedException < Exception
	end

	# A TransitionUndefinedException is thrown by
	# an SMC-generated state machine whenever a transition is taken
	# which:
	#
	# * Is not explicitly defined in the current state.
	# * Is not explicitly defined in the current FSM's default state.
	# * There is no Default transition in the current state.
	class TransitionUndefinedException < Exception
	end

	# base State class
	class State

		def initialize(name, id)
			@_name = name
			@_id = id
		end

		# Returns the state's printable name.
		def getName()
			return @_name
		end

		# Returns the state's unique identifier.
		def getId()
			return @_id
		end

	end

	# The user can derive FSM contexts from this class and interface
	# to them with the methods of this class.
	#
	# The finite state machine needs to be initialized to the starting
	# state of the FSM.  This must be done manually in the constructor
	# of the derived class.
	class FSMContext

		def initialize(initState)
			@_state = initState
			@_previous_state = nil
			@_state_stack = []
			@_transition = nil
			@_debug_flag = nil
			@_debug_stream = $stderr
		end

		# Returns the debug flag's current setting.
		def getDebugFlag()
			return @_debug_flag
		end

		# Sets the debug flag.
		#
		# A true value means debugging is on and false means off.
		def setDebugFlag(flag)
			@_debug_flag = flag
		end

		# Returns the stream to which debug output is written.
		def getDebugStream()
			return @_debug_stream
		end

		# Sets the debug output stream.
		def setDebugStream(stream)
			@_debug_stream = stream
		end

		# Returns the current state.
		def getState()
			if @_state.nil? then
				raise Statemap::StateUndefinedException
			end
			return @_state
		end


		# Is this state machine already inside a transition?
		#
		# True if state is undefined.
		def isInTransition()
			return @_state == nil
		end

		# Returns the current transition's name.
		#
		# Used only for debugging purposes.
		def getTransition()
			return @_transition
		end

		# Clears the current state.
		def clearState()
			@_previous_state = @_state
			@_state = nil
		end

		# Returns the state which a transition left.
		#
		# May be nil.
		def getPreviousState()
			return @_previous_state
		end

		# Sets the current state to the specified state.
		def setState(state)
			raise "undefined state.\n" if state.nil?
			raise "#{state} is not a Statemap.State.\n" unless state.is_a?(Statemap::State)
			@_state = state
			if @_debug_flag then
				@_debug_stream.puts "ENTER STATE     : %s\n" % @_state.getName
			end
		end

		# Returns true if the state stack is empty and false otherwise.
		def isStateStackEmpty()
			return @_state_stack.empty?
		end

		# Returns the state stack's depth.
		def getStateStackDepth()
			return @_state_stack.size
		end

		# Push the current state on top of the state stack
		# and make the specified state the current state.
		def pushState(state)
			raise "undefined state.\n" if state.nil?
			raise "#{state} is not a Statemap.State.\n" unless state.is_a?(Statemap::State)
			unless @_state.nil?
				@_state_stack.push @_state
			end
			@_state = state
			if @_debug_flag then
				@_debug_stream.puts "PUSH TO STATE   : %s\n" % @_state.getName
			end
		end

		# Make the state on top of the state stack the current state.
		def popState()
			if @_state_stack.empty? then
				if @_debug_flag then
					@_debug_stream.puts "POPPING ON EMPTY STATE STACK.\n"
				end
				raise "empty state stack.\n"
			else
				@_state = @_state_stack.pop
				if @_debug_flag then
					@_debug_stream.puts "POP TO STATE    : %s\n" % @_state.getName
				end
			end
		end

		# Remove all states from the state stack.
		def emptyStateStack()
			@_state_stack = []
		end

	end

end
